### PR TITLE
🌱 generate JUnit report in ci-test.sh

### LINKED ISSUE #4400: ci-test.sh should provide JUnit reports  for testgrid
**User Story**

As a developer it would be easier to get an overview over flaky unit test if we would have JUnit reports as an output of ci-test.sh.

**Detailed Description**

For comparison:
* test results ci-e2e.sh: https://testgrid.k8s.io/sig-cluster-lifecycle-cluster-api#capi-e2e-main&width=20
* test results ci-test.sh: https://testgrid.k8s.io/sig-cluster-lifecycle-cluster-api#capi-test-main&width=20

**Anything else you would like to add:**

In case we think this would be useful, I would investigate what the best way is to generate those reports.
Some options:
* [go-junit-report](https://github.com/jstemmer/go-junit-report): create junit reports based on go test output, e.g.: `go test -v 2>&1 | go-junit-report > report.xml`
* [gotestsum](https://github.com/gotestyourself/gotestsum): runs `go test --json`. Seems to provide a very nice output (at a first glance). Could be used via `gotestsum --junitfile unit-tests.xml`
 
I think I would prefer `go-junit-report` because it just converts the go test output to a JUnit file. Shouldn't be a problematic dependency.

/kind feature

### PR DESCRIPTION
**Pull request by sbueringer**
**What this PR does / why we need it**:

Details in the linked issue

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #4400

Relates to #4407. I didn't want to duplicate the dependency into the Docker Makefile. Not sure if that's the correct way to handle this for now, if we want to clean it up in #4407.

Example output:
* https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api/4410/pull-cluster-api-test-main/1376960688349515776
* https://gcsweb.k8s.io/gcs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api/4410/pull-cluster-api-test-main/1376960688349515776/artifacts/
* https://testgrid.k8s.io/sig-cluster-lifecycle-cluster-api#capi-pr-test-main&width=20
  * looks like test grid is about 2h behind

### CODE REVIEW

File: test/infrastructure/docker/Makefile [hunk 1]
```diff
@@ -26,6 +26,8 @@ GOPROXY := https://proxy.golang.org
 endif
 export GOPROXY
 
+REPO_ROOT := $(shell git rev-parse --show-toplevel)
```
[the remark this answers is not in the document]
Comment (sbueringer) [reply test/infrastructure/docker/Makefile]: I already adjusted my code to be compatible with #4412. But I would like to wait with merge anyway.

File: test/infrastructure/docker/Makefile [hunk 2]
```diff
@@ -47,7 +47,7 @@ endif
 # Binaries.
 CONTROLLER_GEN := $(TOOLS_BIN_DIR)/controller-gen
 CONVERSION_GEN := $(TOOLS_BIN_DIR)/conversion-gen
```
Comment (sbueringer) [inline test/infrastructure/docker/Makefile:49]: GOLANGCI_LINT was unused. Probably because the top-level golangci-lint target lints the whole repo

**Comment by sbueringer** [comment]
Okay there's still a bug somewhere. Tests currently always pass even if there are failures.

**Comment by sbueringer** [comment]
@fabriziopandini

 I'm not sure but I'll take a look :)

**Comment by sbueringer** [comment]
> @fabriziopandini
> 
> I'm not sure but I'll take a look :)

@fabriziopandini Currently the XML looks like this:
```xml
	<testsuite tests="3" failures="0" time="0.000" name="sigs.k8s.io/cluster-api/test/infrastructure/docker/api/v1alpha4">
		<properties>
			<property name="go.version" value="go1.16.2"></property>
		</properties>
		<testcase classname="v1alpha4" name="TestDockerMachineTemplateInvalid" time="0.000"></testcase>
time="0.000"></testcase>
                ...
	</testsuite>
```

I couldn't find an easy way to adjust the output of `go test` or `go-junit-report`. 
The package / testsuite name is only dropped in Prow. testgrid shows the full paths. 

Of course we can re-write the XML but I"m not sure if it's worth it.

File: Makefile [hunk 3]
```diff
@@ -126,6 +129,11 @@ test: ## Run tests.
 test-verbose: ## Run tests with verbose settings.
 	TEST_ARGS="$(TEST_ARGS) -v" $(MAKE) test
 
+.PHONY: test-junit
+test-junit: $(GO_JUNIT_REPORT) ## Run tests with verbose setting and generate a junit report.
+	($(MAKE) test-verbose; echo $$? > $(ARTIFACTS)/junit.exitcode) 2>&1 | tee /dev/stderr | $(GO_JUNIT_REPORT) > $(ARTIFACTS)/junit.xml
+	exit $$(cat $(ARTIFACTS)/junit.exitcode)
+
```
Comment (sbueringer) [inline Makefile:136]: This looks quite hacky but it was the simplest way I could achieve the following:
* print the output of go test to the terminal (`tee /dev/stderr`)
* always generate a junit.xml, especially when go test fails
* propagate the exit code, so the make target still fails when go test fails

**Comment by sbueringer** [comment]
/test pull-cluster-api-test-main

(will try the test a few times to find/hit a flaky test once and see if the job fails)

**Comment by sbueringer** [comment]
Test run with failure: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api/4410/pull-cluster-api-test-main/1377128674255441920

I'll squash before merge.

**Comment by sbueringer** [comment]
@fabriziopandini k/k doesn't have this problem because they are using `gotestsum` to generate the reports. I'll try a version with `gotestsum`

**Comment by sbueringer** [comment]
> @fabriziopandini
> 
> Regarding: [#4410 (comment)](https://github.com/kubernetes-sigs/cluster-api/pull/4410#issuecomment-810564924)
> k/k doesn't have this problem because they are using `gotestsum` to generate the reports. I'll try a version with `gotestsum`

Okay so that should be it. Implementation is now very similar to k/k: https://github.com/kubernetes/kubernetes/blob/f981628ce897a96a8a08a2a21c4d6bdb41601fbd/hack/make-rules/test.sh#L249-L256

File: Makefile (hunk 3 above)
Comment (sbueringer) [reply Makefile:137]: In addition: now basically the same implementation as k/k: https://github.com/kubernetes/kubernetes/blob/f981628ce897a96a8a08a2a21c4d6bdb41601fbd/hack/make-rules/test.sh#L249-L256

We now have full package paths both in the Deck junit lens and in testgrid

**Comment by sbueringer** [comment]
Rebased onto master after merge of #4407

I'll break the tests again, just to be absolutely safe.

**Comment by sbueringer** [comment]
> looks great! please squash commits

done

**Review by MarcelMue — APPROVED** [review]
/LGTM

**Review by vincepri — APPROVED** [review]
/approve

**Pull request merged**